src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -1289,70 +1281,69 @@ private static unsafe bool TryInt32ToHexStr(int value, char hexBase, int digits,
             }
 
             charsWritten = bufferLength;
-            fixed (char* buffer = &MemoryMarshal.GetReference(destination))
-            {
-                char* p = Int32ToHexChars(buffer + bufferLength, (uint)value, hexBase, digits);
-                Debug.Assert(p == buffer);
-            }
+
+            int p = Int32ToHexChars(destination, bufferLength, (uint)value, hexBase, digits);
+            Debug.Assert(p == 0);
+
             return true;
         }
 
-        private static unsafe char* Int32ToHexChars(char* buffer, uint value, int hexBase, int digits)
+        private static int Int32ToHexChars(Span<char> buffer, int bufferEndIndex, uint value, int hexBase, int digits)
         {
             while (--digits >= 0 || value != 0)
             {
                 byte digit = (byte)(value & 0xF);
-                *(--buffer) = (char)(digit + (digit < 10 ? (byte)'0' : hexBase));
+                buffer[--bufferEndIndex] = (char)(digit + (digit < 10 ? (byte)'0' : hexBase));
                 value >>= 4;
             }
-            return buffer;
+            return bufferEndIndex;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // called from only one location
-        private static unsafe void UInt32ToNumber(uint value, ref NumberBuffer number)
+        private static void UInt32ToNumber(uint value, ref NumberBuffer number)
         {
             number.DigitsCount = UInt32Precision;
             number.IsNegative = false;
 
-            byte* buffer = number.GetDigitsPointer();
-            byte* p = UInt32ToDecChars(buffer + UInt32Precision, value, 0);
+            Span<byte> buffer = number.Digits;
+            int p = UInt32ToDecChars(buffer, UInt32Precision, value, 0);
 
-            int i = (int)(buffer + UInt32Precision - p);
+            int i = UInt32Precision - p;
 
             number.DigitsCount = i;
             number.Scale = i;
 
-            byte* dst = number.GetDigitsPointer();
+            int dst = 0;
             while (--i >= 0)
-                *dst++ = *p++;
-            *dst = (byte)('\0');
+                buffer[dst++] = buffer[p++];
+            buffer[dst] = (byte)('\0');
 
             number.CheckConsistency();
         }
 
-        internal static unsafe byte* UInt32ToDecChars(byte* bufferEnd, uint value, int digits)
+        internal static int UInt32ToDecChars(Span<byte> buffer, int bufferEndIndex, uint value, int digits)
```

Comment:
:thumbsup: I did a test with Unsafe and it improved performance: https://github.com/dotnet/runtime/pull/38696#issuecomment-659039547

Reply:
I am typically looking at the blast radius of potential bugs. If you omit the bounds check here, how much other code has to be 100% right wrt buffer length computations to avoid silent buffer overrun in release build? If the blast radius is contained within the method (ie bugs in other method won't lead to buffer overrun), I am generally ok with it.

Related - does this method need to take both `buffer` and `bufferEndIndex` ?

Reply:
> Related - does this method need to take both buffer and bufferEndIndex ?

No, `buffer` could be sliced to the correct length.

> I am typically looking at the blast radius of potential bugs. If you omit the bounds check here, how much other code has to be 100% right wrt buffer length computations to avoid silent buffer overrun in release build? If the blast radius is contained within the method (ie bugs in other method won't lead to buffer overrun), I am generally ok with it.

👍. I think in this case, it is self contained, particularly with an assert.

Perhaps though, with it being sliced to the correct length (rather than taking `bufferEndIndex` and/or `digits`) and with a check against `buffer.Length` then for the loop condition, that would be sufficient to get the JIT to elide the checks naturally.

Reply:
hmmm,  looks like the JIT isn't recognizing the reverse iteration: <https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGBLAOwwwo/bABsm5JExQMAKjFwYAFAGUADtn4AeYAE8hAPgbAArgDNzwgJRcenajyd9BfBgF4TFq1AB0AGRh+AHMMAAsGOAZyAG47ZwZ45wB3MN4xGAZlXgZDTwAGW0cEjiSSr0thAG1eAF0PLP0hawByfJa44vLuHLhPWLKeAF8kkeohoA===>

It doesn't do it for arrays either, so possibly something we should fix.

Reply:
What if we use uint? https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGBLAOwwwo/bABsm5JExQMAKjFwYAFAGUADtn4AeYAE8hAPgbAArgDNzwgJRcenajyd9BfBgF4TFq1AB0AGRh+AHMMAAsGOAZyAG47ZwZ45wB3MN4xGAZlXgZDTwAGW0cEjiSSr0thAG1eAF0PLP0hawByfJa44vLuHLhPWLKeAF8kkeokgSERcUlpYlkFJQBVSbVNHSaYYzNKqCKnB3LTSbdPZWPBax2fAKDQiKiBrudB7lT0zOzcgv2Sw+7uNdqtlLnUGspNq12p0AT1Iv0YSUxk5kTwJoJhKIJMQpDJWKYoLBBEsAJKCOikOQQAAiMDALDC2CguDWWl0Bi2FR8aBcGC5wgAovwACZk4UwBC/bj/BInHKeIFQIWikUSyLRREJV4Md4ZLI5PIMQramXdRU1epnSFtDrapy9BHa1HcZ2JZ6TTEzHFzBaKFSbKr1RVS0rPe2ueX8vyBELhdVPcra3WfA0/E12njmsFWjlQ21hkoOjVO0Y0IZAA==

Reply:
That looks to be better codegen and looks to only throw if `buffer.Length == 0` (it shouldn't be) It's probably worth testing and checking codegen in relation to the unsafe version

Reply:
I did some more tests. Using a workaround it is possible to improve the codegen, but it is still bugger than unsafe versions: [Comparison](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mpoDMTUgyEMA3jQYfPH3sahdsktpIDBw+GAwAKqoYABSyYthcADzAAJ4YMAB8DMAcAGZ5MgCU7l6ebtRlVSFhOfmFUACiXAAmRi0wCAwAvAwxoTxFuQUyLAAyMFwA5hgAFgxwDOQaldVeAO6zUjB9ww3Nba2dDJm9dCWrax4VV7d1I1AA2jFhQ/UyB+2dALo9u+kwIoAcjoQJWdwYAF9SrdiAB2e77VpfBDg6rQy4MGFlMIyfyBZjBbi4bCFHIAgBUDAAqkYMLZSBEIAARGBgISzbBQXAAeSgvCmPgCMTSGSpew+rTQNR4DAUAQ4MGltRaAt4GFwFzWNzum22fTgcFVgo1JzODEIhDlCp2AEIzlq7jqIR4BuFYPhsD4OlA0S6+vLJIrpR6vUcoEU/szeAo2DB8DFA8Gluc/S6KTFDRKmq1I70RQCijFQ96ZAwANQMEFAopptYYiHwxGSlp1zwNzzYry4vwBIIMWq0ngMpms9mc7lCDhQWA8OIJZKirLNqDSt3WoNKgey43qzVd8oHqq1bOfI5dXrZ8aTGazNtVPXSA1GtWm04MRiWjeKhj2j+O25nRddcS3De8rkTG0Q3jMMfTzBho1jeNIM3aVyFTI8rmzR4s3eHNDg6BBfnzJci1An0KyrUFa0wrwO1hBFT2Rc9wPorFMW7Xx8X7Qc6RHFk2Q5LlcAAdWgABrLkIG4Fp50SFIAWybM11qJMtxVV99w4w9tOPWUmII45Lzw69pjmcCNi2J8YkzF8TVwM0PwtK01N/B0GAAMg8hgAHoqR5LhJFSBgMAgBgYEkXgOhyaTWgcsBZjZcSGApHy+jdItcIeM9CMjJJ0teK8JjM2YAKuICIRAmDS19WjqhQ5NyOKKMYzjBM1LQjDdKwvDHgMlFiP+DIyOq8NKOrGjurouqPCbfqWMwtjMJ7bjCW3cIh3pRkBPHYTqS4ElCjkxdFJXFTZQ69aGF3DUyq8CrqhPPCcqMldTNvCzPEfHZbJuhz30/FybTc/8ZtcMGvCqz0as+qoGq3JqIxapD2qglNJv9DwfJ82A8hpA7SUEABBFpZNxhgAFl42gVIKeEzlJBYABxGAMDjBpJhUEU8KKaUsqRQyEHggthuLUaKMrCbYbKcmlxXISoD+cn9sO4nSbFvGqfwGm6e5BnmdZ9mZE5mBuYeXmFjgebculrxswVv4RcBMXobGyXqOlti1jm57mMI1jsUwx4qbmCA2nESQYhD2Yw4MCOeTEPgIAOlgiamKZYFwXAYxgIxIq4HwpiKb5MOIew1v6WoAEUOAgPhJgwc7wja2DmsQtrK9laQ8kbmV3QFWYMDunTKtqABHWv69lXpu/CNL+SmQfpabGIJ7r3gG+lWeFj6Nep/CKkF8HooMfRGhISAA===)